### DISCUSSION TITLE
Android hardware keyboard unmapped KeyCodes

### DISCUSSION THREAD

**Discussion by IanRawley** [AvaloniaUI/Avalonia#11275] [Q&A]
I'm in the process of evaluating options for a cross-platform .NET framework for a data entry app we will be developing.  Avalonia is looking like the best option so far due to the ease of access to keyboard events at a page level without other controls.  However, in testing with RawInputEventArgs on an Android device with hardware keyboard (a Juniper Systems Allegro 3) a number of keys are producing KeyUp/KeyDown events with no key or modifiers.  Specifically the digits on the numpad, the alt key, and any combination of the blue function key and alphabet keys.  

In looking into why, I found that a number of native Android keycodes are not mapped to any Avalonia keycodes in [AndroidKeyboardDevice.cs](https://github.com/AvaloniaUI/Avalonia/blob/acbb412a0240c6ad3752afd061f7c3e6c6d7c19a/src/Android/Avalonia.Android/Platform/Input/AndroidKeyboardDevice.cs), much like the following:      
    ` //{ Keycode.?, Key.D0 }`

[AndroidKeyboardEventsHelper.cs](https://github.com/AvaloniaUI/Avalonia/blob/5d785ca99ad8d24bece6ff3f2c3dc0fad367fcde/src/Android/Avalonia.Android/Platform/Specific/Helpers/AndroidKeyboardEventsHelper.cs) also doesn't check for ALT as a modifier, as seen here:

```c#
private static RawInputModifiers GetModifierKeys(KeyEvent e)
        {
            var rv = RawInputModifiers.None;

            if (e.IsCtrlPressed) rv |= RawInputModifiers.Control;
            if (e.IsShiftPressed) rv |= RawInputModifiers.Shift;

            return rv;
        } 
```  
Does anyone know why these keys aren't mapped?  If not, is this something I should look into changing?

**Comment by timunie** [comment]
Mobile is a relative new platform for us to support, so probably these keys were just forgotten. @emmauss what do you think, is it worth a PR? I personally think so.

**Comment by IanRawley** [comment]
I kind of thought that might be the case, but can't be sure without being the one who did it in the first place.  Is there any more discussion, or should I convert this to an issue?

**Reply by maxkatz6** [reply]
Converted and added tags, thanks!

**Reply by IanRawley** [reply]
Thanks.  I'll keep an eye on it.
